src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UnliftedNewtypes #-}
+
+-- |
+-- This module defines a linear wrapper around 'GHC.MutableArray#'
+-- which can be manipulated in pure contexts.
+--
+-- It ensures that effects happen in order. However, access to
+-- out-of-bounds indices causes undefined behaviour.
+--
+-- This module is meant to be imported qualified.
+module Data.Array.Mutable.Unlifted.Linear
+  ( Array#
+  , unArray#
+  , alloc
+  , allocBeside
+  , lseq
+  , size
+  , read
+  , write
+  , copyInto
+  ) where
+
+import Data.Unrestricted.Linear hiding (lseq)
+import Prelude (Int)
+import qualified Prelude as Prelude
+import qualified Unsafe.Linear as Unsafe
+import qualified GHC.Exts as GHC
+
+-- | A mutable array holding @a@s
+newtype Array# a = Array# (GHC.MutableArray# GHC.RealWorld a)
+
+-- | Run an unsafe computation using the underlying MutableArray,
+-- consuming the 'Array#' in process.
+unArray# :: (GHC.MutableArray# GHC.RealWorld a -> b) -> Array# a #-> Ur b
+unArray# f = Unsafe.toLinear (\(Array# a) -> Ur (f a))
+
+-- | Consume an 'Array#'.
+--
+-- Note that we can not implement a 'Consumable' instance because 'Array#'
+-- is unlifted.
+lseq :: Array# a #-> b #-> b
+lseq = Unsafe.toLinear2 (\_ b -> b)
+
+-- | Allocate a mutable array of given size using a default value.
+--
+-- The size should be non-negative.
+alloc :: Int -> a -> (Array# a #-> Ur b) #-> Ur b
+alloc (GHC.I# s) a f =
+  let !new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in f new
```

Comment:
Why is necessary to force `new`? Doesn't the type of `f` force it to be strict?

Also, what prevents `new` from being floated and reused for two different invocations of `alloc`?

Reply:
> Why is necessary to force new? Doesn't the type of f force it to be strict?

You are right, I can remove the bang, I think.

> Also, what prevents new from being floated and reused for two different invocations of alloc?

Wouldn't the fact that it uses bindings from outer scope (`a`, `s`)  prevents it being floated out?

Reply:
> Wouldn't the fact that it uses bindings from outer scope (a, s) prevents it being floated out?
I don't think so.

Suppose cse transforms this `alloc s a (\arr -> alloc s a (f arr))` to
```Haskell
let myAlloc x = alloc s a x
 in myAlloc (\arr -> myAlloc (f arr))
```
Then inline alloc, and float the `new` binding.

`{-# NOINLINE alloc #-}` looks necessary here, plus threading the f parameter as well:
```Haskell
alloc (GHC.I# s) a f0 =
  let (!new, f) = GHC.runRW# Prelude.$ \st ->
        case GHC.newArray# s a st of
          (# _, arr #) -> (Array# arr, f0)
   in f new
```

Reply:
Or NOINLINE + maybe:
```Haskell
alloc (GHC.I# s) a f =
  GHC.runRW# Prelude.$ \st ->
        case GHC.newArray# s a st of
          (# _, arr #) -> f (Array# arr)
```

Reply:
Thank you @facundominguez , I did not know CSE could do that. I will add the `NOINLINE` pragma.